services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy

```diff
@@ -0,0 +1,869 @@
+        northbound.swapFlowPath(flow.id)
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 400
+        exc.responseBodyAsString.to(MessageError).errorMessage ==
+                "Could not swap paths: Flow $flow.id doesn't have protected path"
+
+        and: "Cleanup: revert system to original state"
+        flowHelper.deleteFlow(flow.id)
+    }
+
+    def "Unable to swap paths for a non-existent flow"() {
+        when: "Try to swap path on a non-existent flow"
+        northbound.swapFlowPath(NON_EXISTENT_FLOW_ID)
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 404
+        exc.responseBodyAsString.to(MessageError).errorMessage ==
+                "Could not swap paths: Flow $NON_EXISTENT_FLOW_ID not found"
+    }
+
+    def "Unable to swap paths for an inactive flow"() {
+        given: "Two active neighboring switches with two not overlapping paths at least"
+        def (Switch srcSwitch, Switch dstSwitch) = getNeighboringSwitchPair(2)
+
+        and: "A flow with protected path"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.allocateProtectedPath = true
+        flowHelper.addFlow(flow)
+
+        and: "All alternative paths are unavailable (bring ports down on the source switch)"
+        List<PathNode> broughtDownPorts = []
+        database.getPaths(srcSwitch.dpId, dstSwitch.dpId)*.path.findAll {
+            it != pathHelper.convert(northbound.getFlowPath(flow.id)) &&
+                    it != pathHelper.convert(northbound.getFlowPath(flow.id).protectedPath)
+        }.unique {
+            it.first()
+        }.each { path ->
+            def src = path.first()
+            broughtDownPorts.add(src)
+            northbound.portDown(src.switchId, src.portNo)
+        }
+        Wrappers.wait(WAIT_OFFSET) {
+            assert northbound.getAllLinks().findAll {
+                it.state == IslChangeType.FAILED
+            }.size() == broughtDownPorts.size() * 2
+        }
+
+        when: "Break ISL on the protected/main paths (bring port down) for changing the flow state to down"
+        def flowPathInfo = northbound.getFlowPath(flow.id)
+        def currentPath = pathHelper.convert(flowPathInfo)
+        def currentProtectedPath = pathHelper.convert(flowPathInfo.protectedPath)
+        def protectedIsls = pathHelper.getInvolvedIsls(currentProtectedPath)
+        def currentIsls = pathHelper.getInvolvedIsls(currentPath)
+        northbound.portDown(protectedIsls[0].dstSwitch.dpId, protectedIsls[0].dstPort)
+        northbound.portDown(currentIsls[0].dstSwitch.dpId, currentIsls[0].dstPort)
+
+        then: "Flow state is changed to down"
+        Wrappers.wait(WAIT_OFFSET) { assert northbound.getFlowStatus(flow.id).status == FlowState.DOWN }
+
+        when: "Try to swap the inactive flow"
```

Comment:
I think it's more interesting to swap broken and alive paths rather than try to swap two broken paths